File: Java/src/_0050_Pow_x_n/Solution.java
package _0050_Pow_x_n;

public class Solution {

    public double myPow(double x, int n) {
        return pow(x, n);
    }

    private double pow(double x, long n) {

        if (n < 0) {
            x = 1.0 / x;
            n *= -1;
        } else if (n == 0) return 1;
        else if (n == 1) return x;

        if (n % 2 == 0) {
            double half = pow(x, n / 2);
            return half * half;
        } else return x * pow(x, n - 1);  // n % 2 == 1
    }

    public static void main(String[] args) {

        System.out.println((new Solution()).myPow(2.0, 10));
        System.out.println((new Solution()).myPow(2.1, 3));
        System.out.println((new Solution()).myPow(2.0, -2));
        System.out.println((new Solution()).myPow(2.0, -2147483648));
    }
}

File: Python/_0050_Pow_x_n/Solution.py
# coding=utf-8
# Date: 2021/5/23 21:55
from typing import List


# 执行用时：44 ms, 在所有 Python3 提交中击败了40.22%的用户
# 内存消耗：14.7 MB, 在所有 Python3 提交中击败了92.61%的用户
class Solution:
    def myPow(self, x: float, n: int) -> float:
        is_minus = n < 0
        k = abs(n)
        res = 1
        while k > 0:
            if (k & 1) == 1:
                res *= x
            x *= x
            k >>= 1
        if is_minus:
            res = 1 / res
        return res


if __name__ == "__main__":
    print(Solution().myPow(2.0, 10))
